module/zfs/dbuf.c

```diff
@@ -2701,6 +2789,167 @@ dmu_buf_is_dirty(dmu_buf_t *db_fake, dmu_tx_t *tx)
 	return (dr != NULL);
 }
 
+void
+dmu_buf_direct_mixed_io_wait(dmu_buf_impl_t *db, uint64_t txg, boolean_t read)
+{
+	ASSERT(MUTEX_HELD(&db->db_mtx));
+
+	if (read == B_TRUE) {
+		/*
+		 * If a buffered read is in process, a Direct I/O read will
+		 * wait for the buffered I/O to complete.
+		 */
+		ASSERT3U(txg, ==, 0);
+		while (db->db_state == DB_READ) {
+			DBUF_STAT_BUMP(direct_mixed_io_read_wait);
+			cv_wait(&db->db_changed, &db->db_mtx);
+		}
+	} else {
+		/*
+		 * There must be an ARC buf associated with this Direct I/O
+		 * write otherwise there is no reason to wait for previous
+		 * dirty records to sync out.
+		 *
+		 * The db_state will temporarily be set to DB_CACHED so that
+		 * that any synchronous writes issued through the ZIL will
+		 * still be handled properly. In particular, the call to
+		 * dbuf_read() in dmu_sync_late_arrival() must account for the
+		 * data still being in the ARC. After waiting here for previous
+		 * TXGs to sync out, dmu_write_direct_done() will update the
+		 * db_state.
+		 */
+		ASSERT3P(db->db_buf, !=, NULL);
+		ASSERT3U(txg, >, 0);
+		db->db_mixed_io_dio_wait = TRUE;
+		db->db_state = DB_CACHED;
+		while (dbuf_find_dirty_lte(db, txg) != NULL) {
+			DBUF_STAT_BUMP(direct_mixed_io_write_wait);
+			cv_wait(&db->db_changed, &db->db_mtx);
```

Comment:
So the main factor here was if there were previous dirty records associated with the dbuf, we could not simply just destroy the ARC buffer in case `dr_data` was pointing at the dbuf's `db_buf`. However, I have completely reworked all this in the latest commit `Fixing race condition with rangelocks`. I actually inverted things now by removing the ARC buffer up front before even issuing the Direct I/O write.

This cleans up the code a lot. I actually drew inspiration from the block cloning code with `dmu_buf_will_clone()`. However I do have a question. Is the assumption with block cloning there is no previous dirty records that maybe associated with the dbuf's ARC buffer (`db_buf`)? Is so, then I can see how the ARC buffer could just be destroyed up front. If not though, I feel like there might be an issue there as well. See my latest commit I pointed at above.

The big part of this commit as well, is it removes the race condition we were seeing the CI runs with the `dio_random` test case. By no longer holding the rangelock and going to sleep as a writer, there is no conflicts with `zfs_get_data()` any longer. That also allowed me to basically return all the logic in `zfs_get_data()` to its original state. Let me know your thoughts on this newest commit. Also if you have any idea if there is an issue with the `dmu_buf_will_clone()` logic.

Reply:
I'll look into the updated patch later, but about the questions: If the previous dirty record belongs to a different/earlier TXG, then its dr_data should be separated from the dbuf's buffer once it is dirtied in the current/following TXG.  It is required since the dirty record in process of writing must be immutable.  If the last dirty record belongs to the current TXG, then same as in case of block cloning you should be able to drop the dirty record andcreate a new one you need.  That is what dmu_buf_will_clone() is doing.

Reply:
Sounds good. So I have my concerns that with the Direct I/O write code that that this assumption may not be true. While a dirty record is immutable, this is only possible though `dbuf_fix_old_data()`. Where this happens is when freeing a range, the dbuf is dirtied for the for time, and reading a dbuf. However, Direct I/O happens completely in open-context. There is nothing in that that logic that says that any these cases happen after a Direct I/O write overwrites a buffered (ARC) write. That is the entire reason behind all this logic. I do wonder if that might also be the case for block cloning? Maybe not though? Nonetheless the that is reason for all this code. In the Direct I/O write case there entirely the possibility that the `db_buf` is pointed at by `dr_data` with mixed I/O from an application. If that that is the case, the the immutability thing falls flat as far as I can tell. The only way to make it immutable is through a copy normally done through `dbuf_fix_old_dataIO`. Let me know though if I am missing something. There is nothing in the dbuf code that I have looked at that makes me think this logic is wrong. Happy to be wrong though : ).

Reply:
So I looked at this more today. I am pretty sure a copy needs to be made like I am doing. The invariant part of the dirty record relies on `dbuf_dirty()` -> `dbuf_fix_old_data()` being called when a dirty record is first created for a TXG. Just like in the block cloning case, Direct I/O skips the call to `dbuf_fix_old_data()` because the `db_state` is set to `DB_NOFILL`. Without the call to `dbuf_fix_old_data()`, there is no just in time copy made for the previous TXG dirty record. So, if the previous dirty records `dr_data` is in fact pointed at `db_buf` or `db_data`, then just calling `arc_buf_destroy()`, setting `db_buf = NULL`,  and setting `db_data = NULL` in `dbuf_clear_data()` could lead to a NULL pointer deference in `dbuf_write()` for the previous dirty record when it is being synced out in `dmu_write()`. Let me know though if I am missing something?

Reply:
I think you are heavily overdoing here.  You do not need to copy the buffer if it is shared between the last dirty record and the dbuf.  You just need to delete the pointers from dbuf without freeing the buffer, just leaving it completely to the dirty record.  See what dbuf_fix_old_data() does in the last case if dbuf has no more holds.  Also you do not need to check dirty records before the last, since the should always have their own buffers by this time.

And I suppose after that you should get what dmu_buf_will_clone() should have been. I think it may have an issue now.  I don't see a point to have two different functions.  As I have mentioned before in other context, your code is very close and should better be unified with block cloning.

Reply:
Agreed. I pushed that commit again. Let me know your thoughts. Also, I will be pushing another commit on top of that soon, which will address 90% of your comments so far and further unify this with block cloning.

Reply:
Could you please explain why do you still copy the buffer before destroying?  I've written above how do I see it: just take the last dirty record after undirtying the dbuf for this txg, and compare its buffer with dbuf buffer.  If they are different -- free the dbuf's one, and then in any case wipe the pointers.

Reply:
Happy to update this if we want more mechanics into the code. I figured if `dbuf_fix_old_dataq()` did the job there would be no reason to do this though. My reading of `duf_fix_old_data()` was that it handled the other tow states of a TXG and takes care of chains of updates. I was under the impression though that we wanted to cut down on dbuf  code/logic.I would be on board with that as much as we can.

Reply:
Also, I am happy to look at a code that you think would make the logically/functionally better. Feel free to share a different approach with me.

Reply:
I'd do just this:
```
       if (db->db_buf != NULL) {
                dbuf_dirty_record_t *dr = list_head(&db->db_dirty_records);
                if (dr == NULL || dr->dt.dl.dr_data != db->db_buf)
                        arc_buf_destroy(db->db_buf, db);
                db->db_buf = NULL;
                dbuf_clear_data(db);
        }
```
There is already a dozen of `list_head(&db->db_dirty_records)` calls in the code, so one more or one less...  Especially within dbuf.c I don't see it as problem.

May be could also assert these to make sure the function is not abused:
```
        ASSERT(db->db_level == 0);
        ASSERT(db->db_blkid != DMU_BONUS_BLKID);
        ASSERT(db->db.db_object != DMU_META_DNODE_OBJECT);
```

Reply:
Okay this makes 100% sense to me. Sorry my brain wasn't getting what you were saying without seeing the code lol... My local testing has been going fine with this update. I just pushed it to the commit. Thank you for sharing the code segment.